// Copyright (c) 2014 Baidu.com, Inc. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
//
// Author: qinzuoyan01@baidu.com (Qin Zuoyan)

syntax = "proto3"; 

import "google/protobuf/descriptor.proto";
import "sofa/pbrpc/proto3/rpc_option.proto";

package sofa.pbrpc.builtin;

option cc_generic_services = true;
option java_generic_services = true;
option py_generic_services = true;

message HealthRequest {
}

message HealthResponse {
    string health = 1;
    string version = 2;
    string start_time = 3;
}

message ServerOptions {
    int64 work_thread_num = 1;
    int64 keep_alive_time = 2;
    int64 max_pending_buffer_size = 3;
    int64 max_throughput_in = 4;
    int64 max_throughput_out = 5;
    bool disable_builtin_services = 6;
    bool disable_list_service = 7;
    int64 max_connection_count = 8;
}

message ServerOptionsRequest {
}

message ServerOptionsResponse {
    ServerOptions options = 1;
}

// Update server options.  Only the following options can be update:
//   - keep_alive_time
//   - max_pending_buffer_size
//   - max_throughput_in
//   - max_throughput_out
// If not set, then use old value.
message UpdateOptionsRequest {
    ServerOptions options = 1;
}

// If succeed, returns new values.
message UpdateOptionsResponse {
    ServerOptions options = 1;
}

message ServerStatusRequest {
}

message ServerStatusResponse {
    bool is_listening = 1;

    // count of client connections.
    int64 connection_count = 2;

    // count of registered services.
    int64 service_count = 3;

    // total count of pending messages.
    int64 pending_message_count = 4;

    // total size of pending buffers occupied by pending messages.
    int64 pending_buffer_size = 5;

    // total size of pending data in pending messages, may by less than pending_buffer_size.
    int64 pending_data_size = 6;
}

message ListServiceRequest {
}

message ListServiceResponse {
    // full name of all registered services.
    repeated string services = 1;

    // descriptor of all dependency proto files.
    repeated google.protobuf.FileDescriptorProto files = 2;
}

message MethodStat {
    string method_name = 1;
    int64 succeed_count = 2;
    float succeed_avg_time_us = 3;
    int64 succeed_max_time_us = 4;
    int64 failed_count = 5;
    float failed_avg_time_us = 6;
    int64 failed_max_time_us = 7;
    int64 slot_id = 8;
}

message ServiceStat {
    string service_name = 1;
    int64 period_seconds = 2;
    int64 succeed_count = 3;
    int64 failed_count = 4;
    repeated MethodStat method_stats = 5;
}

message StatRequest {
    string service_name = 1;
    int64 period_seconds = 2; // expect period in seconds
}

message StatResponse {
    repeated ServiceStat service_stats = 1;
    string server_start_time = 2;
}

service BuiltinService {
    option (sofa.pbrpc.service_timeout) = 3000;

    rpc Health(HealthRequest) returns (HealthResponse);

    rpc ServerOptions(ServerOptionsRequest) returns (ServerOptionsResponse);

    rpc UpdateOptions(UpdateOptionsRequest) returns (UpdateOptionsResponse);

    rpc ServerStatus(ServerStatusRequest) returns (ServerStatusResponse);

    rpc ListService(ListServiceRequest) returns (ListServiceResponse) {
        option (sofa.pbrpc.response_compress_type) = CompressTypeGzip;
    }

    rpc Stat(StatRequest) returns (StatResponse);
}
